src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt

```diff
@@ -0,0 +1,179 @@
+            .map { parser.parse(it) }
+            .filter { it.isJsonObject }
+            .mapNotNull { CargoTopMessage.fromJson(it.asJsonObject) }
+            .filter { !it.message.message.startsWith("aborting due to") }
+}
+
+class RsCargoCheckAnnotator : ExternalAnnotator<CargoCheckAnnotationInfo, CargoCheckAnnotationResult>() {
+
+    private fun getCachedResult(file: PsiFile) =
+        CachedValuesManager.getManager(file.project).createCachedValue {
+            CachedValueProvider.Result.create(
+                checkProject(file),
+                PsiModificationTracker.MODIFICATION_COUNT)
+        }
+
+    override fun collectInformation(file: PsiFile, editor: Editor, hasErrors: Boolean): CargoCheckAnnotationInfo? =
+        if (file.project.rustSettings.useCargoCheckAnnotator) {
+            CargoCheckAnnotationInfo(file, editor)
+        } else null
+
+    override fun doAnnotate(info: CargoCheckAnnotationInfo): CargoCheckAnnotationResult? =
+        getCachedResult(info.file).value
+
+    override fun apply(file: PsiFile, annotationResult: CargoCheckAnnotationResult?, holder: AnnotationHolder) {
+        annotationResult ?: return
+        val doc = holder.currentAnnotationSession.file.viewProvider.document ?: throw AssertionError()
+
+        for (topMessage in annotationResult.messages) {
+            val message = topMessage.message
+            val filePath = holder.currentAnnotationSession.file.virtualFile.path
+
+            val severity =
+                when (message.level) {
+                    "error" -> HighlightSeverity.ERROR
+                    "warning" -> HighlightSeverity.WEAK_WARNING
+                    else -> HighlightSeverity.INFORMATION
+                }
+
+            // If spans are empty we add a "global" error
+            if (message.spans.isEmpty()) {
+                if (topMessage.target.src_path == filePath) {
+                    // add a global annotation
+                    val annotation = holder.createAnnotation(severity, TextRange.EMPTY_RANGE, message.message)
+                    annotation.isFileLevelAnnotation = true
+                    annotation.setNeedsUpdateOnTyping(true)
+                    annotation.tooltip = escapeHtml(message.message) + " " + message.code.formatAsLink().orEmpty()
+                }
+            } else {
+                val problemGroup = ProblemGroup { message.message }
+
+                val primarySpan =
+                    message.spans
+                        .filter { it.is_primary }
+                        .filter { isValidSpan(it) }
+                        .filter { filePath.endsWith(PathUtil.getCanonicalPath(it.file_name)) }
+                        .firstOrNull()
+
+                if (primarySpan != null) {
+                    val annotation = createAnnotation(primarySpan, message, severity, doc, holder)
+                    annotation.problemGroup = problemGroup
+                    annotation.setNeedsUpdateOnTyping(true)
+                }
```

Comment:
After the refactoring, maybe `primarySpan == null` is the right condition for global annotations, not `message.spans.isEmpty()`? Otherwise we might lose some messages and don't notice it.